### ISSUE TITLE
Neural Network

### ISSUE THREAD

**Issue by vinzzhang** [vinzzhang/Notes#3]
# 神经网络
输入，权重，激活函数，输出：x，w，g()，output
g(wx + b) = output （b是bias）

#### 激活函数
常用非线性激活函数：sigmoid，tanh，relu；前两者用于**全连接层**，后者relu用于**卷积层**
+ Sigmoid
![image](https://user-images.githubusercontent.com/21994503/44007016-bb116904-9e42-11e8-8f15-1e48fa523027.png)
![image](https://user-images.githubusercontent.com/21994503/44006700-78f53b18-9e3d-11e8-9511-ae44fd7da545.png)

Sigmoid的优点: 便于求导；可以压缩数据，数据幅度不会有太大影响；适合向前传递。

Sigmoid的缺点：容易出现**梯度消失gradient vanishing**，即输入非常大非常小的时候saturation，神经元的梯度（导数）是接近于0的，从图中可以看出梯度的趋势。需要尤其注意参数的初始值来尽量避免saturation的情况； Sigmoid的输出不是**0均值zero-centered**：这会导致后层的神经元的输入是非0均值的信号，这会对梯度产生影响。以 f=sigmoid(wx+b)为例， 假设输入均为正数（或负数），那么对w的导数总是正数（或负数），这样在反向传播过程中要么都往正方向更新，要么都往负方向更新，导致有一种捆绑效果，使得收敛缓慢；**幂运算耗时**。

+ Tanh
![image](https://user-images.githubusercontent.com/21994503/44006954-9ffc0102-9e41-11e8-8a95-2efcdfdde27e.png)
![image](https://user-images.githubusercontent.com/21994503/44006933-519c0606-9e41-11e8-9567-669a6f13d29d.png)

Tanh是sigmoid函数的变形，解决了**non-zero-centered**的问题，其他优缺点与sigmoid函数一样。

+ ReLU
![image](https://user-images.githubusercontent.com/21994503/44007018-c8a90c66-9e42-11e8-8b41-d27903d6325c.png)
![image](https://user-images.githubusercontent.com/21994503/44006938-5effe600-9e41-11e8-888c-8f502e5b58b2.png)

ReLU 的优点：使用 ReLU 得到的SGD的收敛速度会比 sigmoid/tanh 快很多，梯度不会饱和，解决了梯度消失问题；适合用于后向传播。

ReLU 的缺点： **全区间不可导**；训练的时候很”脆弱”：一个非常大的梯度流过一个 ReLU 神经元，更新过参数之后，这个神经元再也不会对任何数据有激活现象了，神经元的梯度就永远都会是0。实际操作中，如果你的**learning rate很大**，那么很有可能你网络中的40%的**神经元dead**；ReLU不会对数据做幅度压缩，所以数据的幅度会随着模型层数的增加不断扩张；ReLU的输出**不是zero-centered**。

一般建议：
+ 少用sigmoid，tanh是sigmoid的优化替代
+ 注意relu的learning rate不要太大（推荐）

#### 神经网络结构及算法
+ 输入层
+ 隐藏层
输入层和隐藏层都存在一个偏置（bias unit) 
  + 学习规则
![image](https://user-images.githubusercontent.com/21994503/44007109-85b3f22a-9e44-11e8-9ed7-8d72d4c3e9ac.png)
![image](https://user-images.githubusercontent.com/21994503/44007112-8b9e28fe-9e44-11e8-80b4-66229f25cfcf.png)
η是learning rate ∈（0,1）
+ 输出层
![image](https://user-images.githubusercontent.com/21994503/44007043-35cc2c92-9e43-11e8-8d14-d2560f373342.png)

#### 多层前馈神经网络
每层神经元与下一层神经元完全互联，神经元之间不存在同层链接也不存在跨层链接。

#### 误差逆转算法BackPropagation，BP
利用输出后的误差来估计输出层的直接前导层的误差，在用这个误差估计更前一层的误差，一层一层的反向传播下去，就获得了所有其他各层的误差估计。

**BP算法**有两个过程：**信号的正向传播和误差的逆向传播**。正向传播最后如果在输出层未得到期望的输出，则转入误差的逆向传播阶段，将输出误差按照某种子形式，通过隐层相输入层返回，并“分摊”给隐层的节点从而获得各层单元的参考误差信号，作为修改各单元权值的依据。这个过程一直持续到误差可以接受或者达到学习次数为止。

**目标函数**：当目标函数（误差函数）小于设定的阈值时即停止神经网络的训练。这里采用最小二乘法：
![image](https://user-images.githubusercontent.com/21994503/44007182-e2763026-9e45-11e8-9caf-c3f4dca9a369.png)
下面以隐层到输出层连接权whj为例进行推导。
BP算法基于梯度下降策略，以目标的负梯度方向对参数进行调整。对误差Ek，给定学习率η有：
![image](https://user-images.githubusercontent.com/21994503/44007225-6643d19c-9e46-11e8-96b8-0cfdf377ed32.png)
注意到whj先影响到到第j个输出层神经元的输入值βj，再影响到其输入值ykj,然后到Ek，有：
![image](https://user-images.githubusercontent.com/21994503/44007238-94e5b916-9e46-11e8-8a7a-562a96245abd.png)
